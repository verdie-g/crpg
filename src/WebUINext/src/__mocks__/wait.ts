export const mockStartMethod = vi.fn();
export const mockEndMethod = vi.fn();
export const mockIsMethod = vi.fn();
export const mockAnyMethod = vi.fn();

const mock = vi.fn().mockImplementation(() => {
  return {
    start: mockStartMethod,
    end: mockEndMethod,
    is: mockIsMethod,
    any: mockAnyMethod,
  };
});

export default mock;
