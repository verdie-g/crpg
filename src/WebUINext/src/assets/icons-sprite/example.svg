<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32"><path d="M26.11 8.63A6 6 0 0 0 21 7.35l-3.43.77-.65.11a5.08 5.08 0 0 1 .82-3.35A2.76 2.76 0 0 1 20 4a1 1 0 0 0 0-2 4.7 4.7 0 0 0-3.78 1.58 6.91 6.91 0 0 0-1.29 4.62l-.51-.08L11 7.35a6 6 0 0 0-5.1 1.28C3.49 10.55 2 14.16 2 18.06c0 4.29 1.72 8.07 4.48 9.85a5.94 5.94 0 0 0 3.25 1 5.93 5.93 0 0 0 2.41-.53 9.29 9.29 0 0 1 7.72 0 5.77 5.77 0 0 0 5.66-.48c2.76-1.77 4.48-5.55 4.48-9.84 0-3.9-1.49-7.51-3.89-9.43Zm-1.68 17.61a3.8 3.8 0 0 1-3.75.32 11.32 11.32 0 0 0-9.36 0 3.8 3.8 0 0 1-3.75-.32C5.4 24.83 4 21.62 4 18.06c0-3.31 1.2-6.32 3.14-7.87a4.18 4.18 0 0 1 2.6-1 3.3 3.3 0 0 1 .81.09l3.43.77a9.18 9.18 0 0 0 4 0l3.43-.77a4 4 0 0 1 3.41.89C26.8 11.74 28 14.75 28 18.06c0 3.56-1.4 6.77-3.57 8.18ZM12 17h-2a1 1 0 0 1-.89-1.45l1-2a1 1 0 0 1 1.78 0l1 2A1 1 0 0 1 12 17Zm10.89-1.45A1 1 0 0 1 22 17h-2a1 1 0 0 1-.89-1.45l1-2a1 1 0 0 1 1.78 0ZM18.5 19h-5a2.5 2.5 0 0 0 0 5h5a2.5 2.5 0 0 0 0-5Zm0 3h-5a.5.5 0 0 1 0-1h5a.5.5 0 0 1 0 1Z"/></svg>
