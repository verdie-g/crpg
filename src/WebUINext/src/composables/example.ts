export const useCounter = () => {
  const counter = ref<number>(0);

  const increase = (): void => {
    counter.value += 1;
  };

  return { counter: readonly(counter), increase };
};
