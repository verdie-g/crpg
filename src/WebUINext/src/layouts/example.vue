<template>
  <main class="px-4 py-10">
    <RouterView />
    <div class="mt-5">[Example Layout]</div>
  </main>
</template>
