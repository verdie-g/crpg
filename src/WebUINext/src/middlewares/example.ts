import type { NavigationGuard } from 'vue-router';

const exampleRouterMiddleware: NavigationGuard = (to, from) => {
  // use return instead next()
  // return { name: 'index' }
  // return '/'
  return true;
};

export default exampleRouterMiddleware;
