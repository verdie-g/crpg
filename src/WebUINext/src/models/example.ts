export interface User {
  name: string;
  role: string;
}
