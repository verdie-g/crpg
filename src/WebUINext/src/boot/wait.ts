// @ts-ignore // TODO: need types defenition, PR to https://github.com/MeForma/vue-wait-for
import wait from '@meforma/vue-wait-for';
import { type InjectionKey } from 'vue';
import { type BootModule } from '@/types/boot-module';

declare class Wait {
  is(name: string): boolean;
  start(name: string): void;
  end(name: string): void;
  any(): number;
}

export const VueWaitKey: InjectionKey<Wait> = Symbol('wait');

export const install: BootModule = app => {
  app.use(wait).provide(VueWaitKey, app.config.globalProperties.$wait);
};
