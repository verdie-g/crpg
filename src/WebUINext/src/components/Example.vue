<script setup lang="ts">
import { useCounter } from '@/composables/example';
import { useUserStore } from '@/stores/example';
import { useWait } from '@/composables/use-wait';

// Provide/Inject
const $w = useWait();

// Props
interface Props {
  msg: string;
}

const props = withDefaults(defineProps<Props>(), {
  msg: 'Prop msg',
});

// Reactive data
const innerMsg = ref<string>('example cmp');
const computedMsg = computed((): string => `${props.msg} - ${innerMsg.value}`);

const waitingKey = 'waiting...';

// Async action with provide/inject calls
const changeInnerMsg = async (): Promise<void> => {
  $w.start(waitingKey);

  await new Promise(resolve => resolve(true));

  innerMsg.value = 'changed example cmp';

  $w.end(waitingKey);
};

// Composable
const { counter, increase } = useCounter();

// Store
const userStore = useUserStore();

// Emit
const emit = defineEmits<{
  (
    e: 'example:changed',
    {
      message,
      superMessage,
    }: {
      message: string;
      superMessage: string;
    }
  ): void;
}>();

// Vue Router
const route = useRoute();
const router = useRouter();

// i18n
const { t } = useI18n();

// TODO: more examples..
// vuelidate
</script>

<template>
  <div>
    <!--  -->
    <div>
      <div>
        prop:
        <span data-aq-example-prop>{{ msg }}</span>
      </div>
      <div>
        ref:
        <span data-aq-example-ref>{{ innerMsg }}</span>
      </div>
      <div>
        computed:
        <span data-aq-example-computed>{{ computedMsg }}</span>
      </div>
      <div>
        emit:
        <button
          data-aq-example-emit
          @click="emit('example:changed', { message: msg, superMessage: computedMsg })"
        >
          Emit, bro
        </button>
      </div>
      <div>
        method + provide/inject:
        <button data-aq-example-method-with-provide-inject @click="changeInnerMsg">
          Method, bro
        </button>
      </div>
      <div data-aq-example-loading-indicator v-if="$w.is(waitingKey)">Loading...</div>
    </div>

    <!--  -->

    <br />
    <br />

    <div>
      <div>
        composable state:
        <span data-aq-example-composable>{{ counter }}</span>
      </div>
      <div>
        composable methods
        <button data-aq-example-composable-method @click="increase">click bro</button>
      </div>
    </div>

    <br />
    <br />

    <!--  -->

    <div>
      <div>
        store action:
        <button data-aq-example-store-action @click="userStore.fetch">fetch user</button>
      </div>
      <div>
        store state:
        <span data-aq-example-store-state>{{ userStore.name }}</span>
      </div>
      <div>
        store getter:
        <span data-aq-example-store-getter>
          {{ userStore.namePlusRole }}
        </span>
      </div>
    </div>

    <br />
    <br />

    <!--  -->

    <div>
      router
      <div>
        route path
        <span data-aq-example-route-path>
          {{ route.path }}
        </span>
      </div>
      <div>
        router methods
        <button data-aq-example-router-push @click="router.push({ name: 'index' })">
          push router, bro
        </button>
      </div>
    </div>

    <br />
    <br />

    <!--  -->

    <div>
      i18n
      <span data-aq-example-i18n>
        {{ t('button.back') }}
      </span>
    </div>

    <br />
    <br />

    <!--  -->

    <div>
      children component
      <SvgIcon
        data-aq-example-children-component
        name="example"
        viewBox="0 0 32 32"
        class="w-12"
      ></SvgIcon>
    </div>
  </div>
</template>
