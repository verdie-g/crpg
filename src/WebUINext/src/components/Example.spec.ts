import { mount } from '@vue/test-utils';
import { createTestingPinia, TestingOptions } from '@pinia/testing';
import { useUserStore } from '@/stores/example';
import { VueWaitKey } from '@/boot/wait';
import MockWait from '@/__mocks__/wait';

import Example from './Example.vue';

// START MOCKS
const mockedWait = new MockWait();

const TEST_PATH = '/test-path';
const mockedPush = vi.fn();

// FIXME: https://github.com/vuejs/router/issues/1466
vi.mock('vue-router/dist/vue-router.mjs', () => ({
  useRoute: vi.fn().mockImplementation(() => ({
    path: TEST_PATH,
  })),
  useRouter: vi.fn().mockImplementation(() => ({
    push: mockedPush,
  })),
}));
// END MOCKS

const factory = (options?: TestingOptions) => {
  const wrapper = mount(Example, {
    props: {
      msg: 'Applejack',
    },
    shallow: true,
    global: {
      provide: {
        [VueWaitKey as symbol]: mockedWait,
      },
      plugins: [createTestingPinia(options)],
    },
  });

  const store = useUserStore();
  return { wrapper, store };
};

describe('Example component unit test', () => {
  it('basic', () => {
    const { wrapper } = factory();

    expect(wrapper.find('[data-aq-example-prop]').text()).toEqual('Applejack');
    expect(wrapper.find('[data-aq-example-ref]').text()).toEqual('example cmp');
    expect(wrapper.find('[data-aq-example-computed]').text()).toEqual('Applejack - example cmp');
  });

  // https://test-utils.vuejs.org/guide/essentials/event-handling.html
  it('emit', async () => {
    const { wrapper } = factory();

    await wrapper.find('[data-aq-example-emit]').trigger('click');

    expect(wrapper.emitted('example:changed')).toHaveLength(1);
    expect(wrapper.emitted('example:changed')?.[0]).toEqual([
      {
        message: 'Applejack',
        superMessage: 'Applejack - example cmp',
      },
    ]);
  });

  it('provide/inject', async () => {
    const { wrapper } = factory();

    await wrapper.find('[data-aq-example-method-with-provide-inject]').trigger('click');

    expect(wrapper.find('[data-aq-example-ref]').text()).toEqual('changed example cmp');

    expect(mockedWait.start).toHaveBeenCalled();
    expect(mockedWait.end).toHaveBeenCalled();
    expect(mockedWait.is).toHaveBeenCalled();
  });

  // https://test-utils.vuejs.org/guide/advanced/reusability-composition.html
  it('with composable', async () => {
    const { wrapper } = factory();

    const composableNodeEl = wrapper.find('[data-aq-example-composable]');

    expect(composableNodeEl.text()).toEqual('0');

    await wrapper.find('[data-aq-example-composable-method]').trigger('click');

    expect(composableNodeEl.text()).toEqual('1');
  });

  // https://pinia.vuejs.org/cookbook/testing.html#unit-testing-components
  it('with store', async () => {
    const { wrapper, store } = factory({
      initialState: { user: { name: 'Twilight Sparkle', role: 'Royalty' } },
    });

    const storeStateNodeEl = wrapper.find('[data-aq-example-store-state]');
    const storeGetterNodeEl = wrapper.find('[data-aq-example-store-getter]');

    expect(storeStateNodeEl.text()).toEqual('Twilight Sparkle');
    expect(storeGetterNodeEl.text()).toEqual('Twilight Sparkle Royalty');

    await wrapper.find('[data-aq-example-store-action]').trigger('click');

    expect(store.fetch).toHaveBeenCalledTimes(1);
  });

  // https://test-utils.vuejs.org/guide/advanced/vue-router.html#using-a-mocked-router-with-composition-api
  it('with route', () => {
    const { wrapper } = factory();

    expect(wrapper.find('[data-aq-example-route-path]').text()).toEqual(TEST_PATH);
  });

  it('with router', async () => {
    const { wrapper } = factory();

    await wrapper.find('[data-aq-example-router-push]').trigger('click');
    expect(mockedPush).toHaveBeenCalledTimes(1);
    expect(mockedPush).toBeCalledWith({ name: 'index' });
  });

  // https://stackoverflow.com/questions/72260793/best-way-to-mock-stub-vue-i18n-translations-in-a-vue3-component-when-using-vites/73630072#73630072
  it('with i18n', async () => {
    const { wrapper } = factory();

    expect(wrapper.find('[data-aq-example-i18n]').text()).toEqual('button.back');
  });

  // https://vitest.dev/guide/snapshot.html
  it('snapshot', async () => {
    const { wrapper } = factory();

    // for update snapshot, use `-- -u`  ex: npm run test:unit src/components/ -- -u
    expect(wrapper.html()).toMatchSnapshot();
  });
});
