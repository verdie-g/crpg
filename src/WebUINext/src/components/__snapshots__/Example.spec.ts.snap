// Vitest Snapshot v1

exports[`Example component unit test > snapshot 1`] = `
"<div>
  <!--  -->
  <div>
    <div> prop: <span data-aq-example-prop=\\"\\">Applejack</span></div>
    <div> ref: <span data-aq-example-ref=\\"\\">example cmp</span></div>
    <div> computed: <span data-aq-example-computed=\\"\\">Applejack - example cmp</span></div>
    <div> emit: <button data-aq-example-emit=\\"\\"> Emit, bro </button></div>
    <div> method + provide/inject: <button data-aq-example-method-with-provide-inject=\\"\\"> Method, bro </button></div>
    <!--v-if-->
  </div><!--  --><br><br>
  <div>
    <div> composable state: <span data-aq-example-composable=\\"\\">0</span></div>
    <div> composable methods <button data-aq-example-composable-method=\\"\\">click bro</button></div>
  </div><br><br><!--  -->
  <div>
    <div> store action: <button data-aq-example-store-action=\\"\\">fetch user</button></div>
    <div> store state: <span data-aq-example-store-state=\\"\\"></span></div>
    <div> store getter: <span data-aq-example-store-getter=\\"\\"> </span></div>
  </div><br><br><!--  -->
  <div> router <div> route path <span data-aq-example-route-path=\\"\\">/test-path</span></div>
    <div> router methods <button data-aq-example-router-push=\\"\\"> push router, bro </button></div>
  </div><br><br><!--  -->
  <div> i18n <span data-aq-example-i18n=\\"\\">button.back</span></div><br><br>
</div>"
`;

exports[`Example component unit testing > snapshot 1`] = `
"<div>
  <!--  -->
  <div> prop: <span data-aq-example-prop=\\"\\">Applejack</span></div>
  <div> ref: <span data-aq-example-ref=\\"\\">example cmp</span></div>
  <div> computed: <span data-aq-example-computed=\\"\\">Applejack - example cmp</span></div>
  <div> emit: <button data-aq-example-emit=\\"\\"> Emit, bro </button></div><!--  -->
  <div> composable state: <span data-aq-example-composable=\\"\\">0</span></div>
  <div> composable methods <button data-aq-example-composable-method=\\"\\">click bro</button></div><!--  -->
  <div> store action: <button data-aq-example-store-action=\\"\\">fetch user</button></div>
  <div> store state: <span data-aq-example-store-state=\\"\\"></span></div>
  <div> store getter: <span data-aq-example-store-getter=\\"\\"> </span></div>
</div>"
`;
